### PR TITLE
Refactor WatermarkFactory::create to WatermarkFactory::createBuffer.

### PR DESCRIPTION
**Pull request by KBaichoo**
Refactor WatermarkFactory::create to WatermarkFactory::createBuffer. we'll be creating BufferMemoryAccounts from the factories as well.

Signed-off-by: Kevin Baichoo <kbaichoo@google.com>

Commit Message: Refactor WatermarkFactory::create to WatermarkFactory::createBuffer.
Additional Description: See https://github.com/envoyproxy/envoy/pull/17093#discussion_r656632550
Risk Level: low 
Testing: Ran tests
Docs Changes: NA
Release Notes: NA
Platform Specific Features: NA
Related Issue #15791

### CODE REVIEW

File: envoy/buffer/buffer.h [hunk 1]
```diff
@@ -480,7 +480,6 @@ class Instance {
 using InstancePtr = std::unique_ptr<Instance>;
 
 /**
- * A factory for creating buffers which call callbacks when reaching high and low watermarks.
```
Comment (antoniovicente) [inline envoy/buffer/buffer.h:483]: I think you should keep this class comment

**Review by antoniovicente — COMMENTED** [review]
Looks good otherwise.

/wait

File: envoy/buffer/buffer.h (hunk 1 above)
Comment (KBaichoo) [reply envoy/buffer/buffer.h]: Whoops, done!

**Review by antoniovicente — APPROVED** [review]

**Pull request merged**